### PR TITLE
[WIP] Minimax with tuple

### PR DESCRIPTION
**Pull request by sethpthomas91**

### CODE REVIEW

File: lib/game.ex [hunk 1]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+
+  def handle_next_move(game) do
+    move = next_open(game)
+    player_move_list = determine_move_list(game)
+    game = update_available_moves(move, game)
+    game = update_player_move_list(game, player_move_list, move)
+    place_marker(move, game)
+  end
+
+  def best_move(game) do
```
Comment (sethpthomas91) [inline lib/game.ex:142]: This is the current iteration when I had an issue with trying to update the best_score outside of the map function.

File: lib/game.ex [hunk 2]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+    {:ok, {best_move, _score}} = Enum.fetch(move_list, index_of_max_score)
+    best_move
+
+  end
+
+  # def minimax(game) do
+
+  # end
+
+  # def minimax(game) do
```
Comment (sethpthomas91) [inline lib/game.ex:189]: Old minimax

File: test/game_test.exs [hunk 3]
```diff
@@ -266,4 +270,217 @@ defmodule Gametest do
...
+      :win? => true,
+      :maximizer => 1,
+      :winner => 2
+    }
+
+    assert Game.score(game) == -10
+  end
+
+  test "best_move returns final move that wins" do
+    game = %{
```
Comment (martingaston) [inline test/game_test.exs:394]: this setup is getting very complex and makes your tests hard to follow (at least using my eyes & brain)

File: test/game_test.exs [hunk 4]
```diff
@@ -266,4 +270,217 @@ defmodule Gametest do
...
+        7 => "7",
+        8 => "X",
+        9 => "O"
+      }
+    }
+
+    assert Game.best_move(game) == 7
+  end
+
+  test "best_move returns final move that blocks a win" do
```
Comment (martingaston) [inline test/game_test.exs:422]: this naming is a little unclear to me here as well, at least without resorting to guesswork/gut feelings.

File: test/game_test.exs [hunk 5]
```diff
@@ -266,4 +270,217 @@ defmodule Gametest do
...
+  test "score should return 0 when the game is a draw" do
+    game = %{
+      :available_moves => [],
+      :win? => false
+    }
+
+    assert Game.score(game) == 0
+  end
+
+  test "score should return 10 when the game is won by maximizer" do
```
Comment (martingaston) [inline test/game_test.exs:373]: should the internals of minimax be exposed here?

File: lib/game.ex [hunk 6]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+
+  def best_move(game) do
+    # best_score = -100
+    # new_move = nil
+    game = assign_maximizer(game)
+
+    # Iterate through every cell
+    score_move_list =
+      Enum.map([1, 2, 3, 4, 5, 6, 7, 8, 9], fn move ->
+        IO.puts("\r\n\r\nTrying == #{move} ")
```
Comment (martingaston) [inline lib/game.ex:150]: i think you are going to reach the limits of printf debugging pretty quickly here - have you tried using the debugger at any point during the apprenticeship?

File: lib/game.ex [hunk 7]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+        IO.inspect(move in game.available_moves)
+        IO.puts("MOVE IN AVAIL MOVES?")
+
+        # If move is in available moves continue
+        if move in game.available_moves do
+          # IO.puts("BEST SCORE BEFORE ==== #{best_score}")
+          IO.puts("DOING == #{move} ")
+          # play the next available move
+          game = handle_move(move, game)
+          game = check_for_win(game)
```
Comment (martingaston) [inline lib/game.ex:167]:
```elixir
score = handle_move(move, game)
|> check_for_win()
|> score()

{move, score}
```

File: lib/game.ex [hunk 8]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+        IO.puts("AVAIL MOVES BELOW")
+        IO.inspect(game.available_moves)
+        IO.puts("AVAIL MOVES ABOVE")
+
+        IO.puts("MOVE IN AVAIL MOVES?")
+        IO.inspect(move in game.available_moves)
+        IO.puts("MOVE IN AVAIL MOVES?")
+
+        # If move is in available moves continue
+        if move in game.available_moves do
```
Comment (martingaston) [inline lib/game.ex:162]: why not just map through `game.available_moves` to begin with?

File: lib/game.ex [hunk 9]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+    IO.puts("WINNER!")
+    10
+  end
+
+  def score(%{win?: true}) do
+    IO.puts("LOSER!")
+    -10
+  end
+
+  def score(game), do: game
```
Comment (martingaston) [inline lib/game.ex:254]: is this a totally separate return type? the other `score` fns all return an int, no?

File: lib/game.ex [hunk 10]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+        # If move is in available moves continue
+        if move in game.available_moves do
+          # IO.puts("BEST SCORE BEFORE ==== #{best_score}")
+          IO.puts("DOING == #{move} ")
+          # play the next available move
+          game = handle_move(move, game)
+          game = check_for_win(game)
+          {move, score(game)}
+
+        else
```
Comment (martingaston) [inline lib/game.ex:170]: by skip you mean return `nil` for your map? 

```elixir
res = Enum.map([0,1,2,3,4,5], fn x ->
  if x in [0,2,4] do
    x
  end
end)

IO.inspect(res) # [0, nil, 2, nil, 4, nil]
```

File: lib/game.ex [hunk 11]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+          # If move is not in available moves skip
+          IO.puts("SKIPPING == #{move}\n")
+        end
+      end)
+
+    move_list = Enum.filter(score_move_list, fn elem -> elem != :ok end)
+    list_of_scores = Enum.map(move_list, fn {_move, score} -> score end)
+    max_score = Enum.max(list_of_scores)
+    index_of_max_score = Enum.find_index(move_list, fn {_move, score} -> score == max_score end)
```
Comment (martingaston) [inline lib/game.ex:179]: this feels very imperative to me 😄

File: lib/game.ex [hunk 12]
```diff
@@ -124,4 +128,128 @@ defmodule Game do
...
+          game = check_for_win(game)
+          {move, score(game)}
+
+        else
+          # If move is not in available moves skip
+          IO.puts("SKIPPING == #{move}\n")
+        end
+      end)
+
+    move_list = Enum.filter(score_move_list, fn elem -> elem != :ok end)
```
Comment (martingaston) [inline lib/game.ex:176]: i'm a little lost, but does this ever return `false` for any element in your `move_list`? `move_list` is a list of `{move, score}` tuples || `nil`, right?

it also feels a little odd in this context to be using this idiom to specifically filter on things that are _not_ `:ok` when we're on the happy/success path.

File: test/game_test.exs (hunk 3 above)
Comment (sethpthomas91) [reply test/game_test.exs:394]: I started changing the test names to help with the confusion. It specifies exactly what the move should be now

File: test/game_test.exs (hunk 5 above)
Comment (sethpthomas91) [reply test/game_test.exs:373]: I think it should be. This really is not the "minimax" algo. This is just making sure when we hit our base cases the scoring system at the end of minimax is good to go.

File: lib/game.ex (hunk 6 above)
Comment (sethpthomas91) [reply lib/game.ex:150]: Yep! Used it in Java, JS, Ruby, and Python. This was after a session with another apprentice where we could both see what was going on with the data visually. It is easy enough to just get rid of those lines quickly.

File: lib/game.ex (hunk 9 above)
Comment (sethpthomas91) [reply lib/game.ex:254]: My intent was to have the game fall through the functions until it became an integer. Once it becomes an integer then we can grab it as the score. Still trying to work this one out.

File: lib/game.ex (hunk 10 above)
Comment (sethpthomas91) [reply lib/game.ex:170]: I have a filter, but I was having an issue in just skipping these. This is a syntax issue that I was fighting.

File: lib/game.ex (hunk 11 above)
Comment (sethpthomas91) [reply lib/game.ex:179]: It is insanely ugly and imperative. I hate it, but what recommendations do you have to try to make it not imperative? Any guidance on things to try to get me out of the imperative thinking here? 

Not really looking for an answer, but suggestions on ways of thinking would be great! Trying to break old habits of OOP.

File: lib/game.ex (hunk 12 above)
Comment (sethpthomas91) [reply lib/game.ex:176]: This is a syntax issue. If I could just not return the :ok s it would be a bit better and no filter would be needed. 

At the end of this I started thinking that I could store the best_move and best_score in the game object itself. Once I do that I may be able to just let this game object fall through until all combos are done, then just grab the best_move key/ value in the object. 

@new_game %{
    :available_moves => [1, 2, 3, 4, 5, 6, 7, 8, 9],
    :win? => false,
    :winner => nil,
    :player_1_moves => [],
    :player_2_moves => [],
    :player_1_mark => "X",
    :player_2_mark => "O",
    :current_player => 1,
    :player_1_type => :human,
    :player_2_type => :human,
    :maximizer => nil,
    :best_score_move => %{
      :best_score => -100,
      :best_move => nil
    },
    :board => %{
      1 => "1",
      2 => "2",
      3 => "3",
      4 => "4",
      5 => "5",
      6 => "6",
      7 => "7",
      8 => "8",
      9 => "9"
    }
  }

File: lib/game.ex (hunk 9 above)
Comment (martingaston) [reply lib/game.ex:254]: ah, i see. an interesting idea! i'm probably too used to static languages to be useful here 😄

File: lib/game.ex (hunk 11 above)
Comment (martingaston) [reply lib/game.ex:179]: so making something pipeable is usually a good start, and losing the trail of variable assignments. that way you're composing functions together, which kind of forces you to write declaratively out of the box.

a simplified of what you're trying to do here would be:

```elixir
defmodule FP do
  def get_best_move(moves, scores) do
    {_, move} = Enum.zip(scores, moves)
    |> Enum.max()

    move
  end
end

moves = [1,2,3,4,5]
scores = [0,0,10,0,-10]

FP.get_best_move(moves, scores)
|> IO.inspect() # 3
``` 

notice we're relying on functions to get everything we need, as opposed to imperatively building everything into a variable, finding the index, then combining everything together.
